### ISSUE TITLE
Add db support to enumerate GUIDs on a host

### ISSUE THREAD

**Issue by amoss** [netdata/netdata#8919]
[Not yet groomed - will be refined]

## In the case that the slave disconnected / reconnected:
* Inside `rrdpush_receive` we retrieve the `RRDHOST *host`.
* Via `host->rrdset_root ... -> next` we can enumerate the charts.
* Via `chart->dimensions` we can enumerate the `RRDDIM`.

Via in the in-memory structures we have access to the chart / dimension identities, and the `last_collected_time`. We can estimate the size of the gap directly from this value, although we will attempt to handle clock-slew and the slave timestamps.

## In the case that the master disconnected / reconnected:
* We do not have this in-memory structure built.
This gives us two options for what to do:
1. Add a query to the beginning of the streaming protocol, this would supply a list of charts / dimensions / `last_collected_time`s from the slave.
2. Rely on the persistent metadata reload to provide these structures.

**Comment by mfundul** [comment]
This is overly complicated and does not seem to cover all cases.

Imagine a collector in the slave that initialized after the streaming sync is done. Then no gap is filled ever.

I propose syncing every metric individually as it begins being streamed by the slave. This covers both slave disconnection and master disconnection. Moreover, it also covers the case of collectors coming online later.

**Comment by thiagoftsm** [comment]
> Add a query to the beginning of the streaming protocol, this would supply a list of charts / dimensions / last_collected_times from the slave.

Only to confirm, this change will happens after handshake that determines among other things the protocol version. Is this conclusion correct?

**Comment by amoss** [comment]
> This is overly complicated and does not seem to cover all cases.
> 
> Imagine a collector in the slave that initialized after the streaming sync is done. Then no gap is filled ever.
> 
> I propose syncing every metric individually as it begins being streamed by the slave. This covers both slave disconnection and master disconnection. Moreover, it also covers the case of collectors coming online later.

This is not less complex, it just moves problems to another place.

1. If we decide to sync when the collector comes online then we have an incoming stream of live data - how should we handle this in parallel to filling the gap?
2. The initialization phase of the stream connection is bi-directional, then it drops to a uni-directional stream. If we sync each metric individually then we need to rewrite streaming so that we are watching for commands from the master to the slave during the normal operation of the link. This adds an extra thread to read the socket and an extra layer of coordination / synchronization in the code.

**Comment by mfundul** [comment]
It's true, it's also complex.

1. I was thinking:
   - If master/slave is not synced live data are ignored.
   - DB latest timestamps are compared and the diff is streamed at the DB level and not the collector level.
   - When sync is finished live data stop being ignored.
2. Yes, a new layer would be required.

**Comment by amoss** [comment]
I don't have a strong understanding yet of which option would be more complex so I'll prepare a description of approaches for the groom on Monday and we can try and estimate complexity then.

**Comment by amoss** [comment]
Closing in favour of new issues (design shifted, we won't need to enumerate GUIDs)
